File: pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

// Prototypes
int get_card_digits(long credit);
int calculate_checksum(long credit);
int get_first_digits(long credit, int n);

// Global variables
const int AMEX = 15;
const int MASTERCARD = 16;
const int VISA1 = 13;
const int VISA2 = 16;

int main(void)
{
    // Prompt user for credit card number
    long card = get_long("Number: ");

    // Get number of card's digits
    int digits = get_card_digits(card);

    // Calculate checksum formula
    int checksum_value = calculate_checksum(card);

    // Check card type
    if (checksum_value != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        int first_two_digits = get_first_digits(card, digits - 2);

        // check if AMEX
        if ((first_two_digits == 34 || first_two_digits == 37) && digits == AMEX)
        {
            printf("AMEX\n");
        }
        // Check if mastercard
        else if ((first_two_digits >= 51 && first_two_digits <= 55) && digits == MASTERCARD) 
        {
            printf("MASTERCARD\n");
        }
        // Check Visa
        else if ((first_two_digits / 10 == 4) && (digits == VISA1 || digits == VISA2))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

// Get digits number
int get_card_digits(long credit)
{
    int digits = 0;

    while (credit > 0)
    {
        digits++;
        credit /= 10;
    }
    return digits;
}

// Calculate checksum
int calculate_checksum(long credit)
{
    int last2second = 0, last = 0;
    bool is_second_to_last = false;

    while (credit > 0)
    {
        int last_digit = credit % 10;

        // Get second to last digit
        if (is_second_to_last)
        {
            // Calculate the product
            last_digit *= 2;

            // has two digits
            if (last_digit > 9)
            {
                last2second += (last_digit % 10);
                last2second += (last_digit / 10);
            }
            // one digit
            else
            {
                last2second += last_digit;
            }
        }
        // Get last digit sum
        else
        {
            last += last_digit;
        }
        
        // Switch label
        is_second_to_last = !is_second_to_last;

        // update card last digit
        credit /= 10;
    }
    return (last2second + last) % 10;
}

// Get first digits
int get_first_digits(long credit, int n)
{
    return (credit / pow(10, n));
}

File: pset6/sentimental-credit/credit.py
# Global
AMEX = 15
MASTERCARD = 16
VISA = [13, 16]


def main():
    # Prompt user for credit card number
    card = input("Number: ")

    # Not digits
    if not card.isdigit():
        print("INVALID")
        return

    # Card length
    ln = len(card)

    # First two digits
    first_digits = int(card[0:2])

    # Calculate checksum
    card = int(card)
    checksum = calculate_checksum(card)

    # Check card type and print it
    if is_AMEX(checksum, first_digits, ln):
        print("AMEX")
    elif is_mastercard(checksum, first_digits, ln):
        print("MASTERCARD")
    elif is_visa(checksum, first_digits, ln):
        print("VISA")
    else:
        print("INVALID")


def calculate_checksum(card):
    last2second = 0
    last = 0
    is_second_to_last = False

    while card > 0:
        # Get last recent digit
        last_digit = card % 10

        # Get second to last digit
        if is_second_to_last:
            last_digit *= 2
            # Calculate the product of 2tolast
            if last_digit > 9:
                last2second += (last_digit % 10)
                last2second += (last_digit // 10)
            else:
                last2second += last_digit
        # Get last digits total
        else:
            last += last_digit
        
        is_second_to_last = not is_second_to_last
        card //= 10
    
    return (last2second + last) % 10


# Check AMEX
def is_AMEX(checksum, first_digits, ln):
    return (checksum == 0) and (first_digits in [34, 37]) and (ln == AMEX)


# Check MASTERCARD
def is_mastercard(checksum, first_digits, ln):
    return (checksum == 0) and (first_digits >= 51 and first_digits <= 55) and (ln == MASTERCARD)


# Check VISA
def is_visa(checksum, first_digits, ln):
    return (checksum == 0) and (first_digits // 10 == 4) and (ln in VISA)


if __name__ == "__main__":
    main()